pkg/awsutil/config.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+	liberrors "github.com/ekristen/libnuke/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+func (c *Credentials) NewConfig(ctx context.Context, region, serviceType string) (*aws.Config, error) {
+	log.Debugf("creating new config in %s for %s", region, serviceType)
+
+	if region == GlobalRegionID {
+		region = "us-east-1"
+	}
+
+	var cfg *aws.Config
+	if customRegion := c.CustomEndpoints.GetRegion(region); customRegion != nil {
+		var opts []func(*config.LoadOptions) error
+
+		customService := customRegion.Services.GetService(serviceType)
+		if customService == nil {
+			return nil, liberrors.ErrSkipRequest(fmt.Sprintf(
+				".service '%s' is not available in region '%s'",
+				serviceType, region))
+		}
+
+		opts = append(opts, config.WithRegion(region))
+		opts = append(opts, config.WithCredentialsProvider(c.awsNewStaticCredentialsV2()))
+		// FUTURE: SDK v2 does not support sideloading of a base endpoint URL
+		//
+		// the only way to force an endpoint from config is to use the
+		// deprecated "v1" endpoint resolver (in the v2 SDK)
+		//
+		// https://github.com/aws/aws-sdk-go-v2/issues/2422
+		opts = append(opts, config.WithEndpointResolver(baseEndpoint{customService.URL}))
+
+		if customService.TLSInsecureSkipVerify {
+			client := &http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, //nolint:gosec
+				},
+			}
+			opts = append(opts, config.WithHTTPClient(client))
+		}
+
+		cfgv, err := config.LoadDefaultConfig(ctx, opts...)
+		if err != nil {
+			return nil, err
+		}
+
+		cfg = &cfgv
+	}
+
+	if cfg == nil {
+		root, err := c.rootConfig(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		cfgCopy := root.Copy()
+		cfgCopy.Region = region
+		cfg = &cfgCopy
+	}
+
+	return cfg, nil
+}
+
+func (c *Credentials) rootConfig(ctx context.Context) (*aws.Config, error) {
+	if c.cfg != nil {
+		return c.cfg, nil
+	}
+
+	var opts []func(*config.LoadOptions) error
+	opts = append(opts, config.WithAPIOptions([]func(*middleware.Stack) error{
+		func(stack *middleware.Stack) error {
+			if err := stack.Finalize.Add(traceRequest{}, middleware.After); err != nil {
+				return err
+			}
+			return stack.Deserialize.Add(traceResponse{}, middleware.After)
+		},
+	}))
+
+	region := DefaultRegionID
+	log.Debugf("creating new root session in %s", region)
+
+	switch {
+	case c.HasAwsCredentials(): // adapts from v1 credentials provider
+		creds, err := c.Credentials.GetWithContext(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		static := credentials.NewStaticCredentialsProvider(
+			creds.AccessKeyID,
+			creds.SecretAccessKey,
+			creds.SessionToken,
+		)
+		opts = append(opts, config.WithCredentialsProvider(static))
+
+	case c.HasProfile() && c.HasKeys():
+		return nil, fmt.Errorf("you have to specify a profile or credentials for at least one region")
```

Comment:
nit: might be better to say "can not specify both profile and credentials for one region" considering it is an error message to correct user config behavior

Reply:
This string was taken from the v1 version of this so I'm going to leave as-is.